pretyping/pretyping.mli

```diff
@@ -130,3 +110,45 @@ val ise_pretype_gen :
 val register_constr_interp0 :
   ('r, 'g, 't) Genarg.genarg_type ->
     (unbound_ltac_var_map -> env -> evar_map -> types -> 'g -> constr * evar_map) -> unit
+
+(** Open-recursion style pretyper *)
+
+open Misctypes
+open Glob_term
+
+module ExtraEnv :
+sig
+  type t
+  val env : t -> Environ.env
+  val make_env : Environ.env -> evar_map -> t
+  val push_rel : evar_map -> rel_declaration -> t -> t
+end
+(** A wrapper around [Environ.env] to generate more efficiently named contexts
+    of fresh evars. *)
+
+type 'a pretype_fun = Loc.t option -> bool -> type_constraint -> ExtraEnv.t -> evar_map ref -> (int * ltac_var_map) -> 'a
+
+type pretyper = {
+  pretype_ref : pretyper -> Globnames.global_reference * glob_level list option -> unsafe_judgment pretype_fun;
+  pretype_var : pretyper -> Id.t -> unsafe_judgment pretype_fun;
+  pretype_evar : pretyper -> existential_name * (Id.t * glob_constr) list -> unsafe_judgment pretype_fun;
+  pretype_patvar : pretyper -> Evar_kinds.matching_var_kind -> unsafe_judgment pretype_fun;
+  pretype_app : pretyper -> glob_constr * glob_constr list -> unsafe_judgment pretype_fun;
+  pretype_lambda : pretyper -> Name.t * Decl_kinds.binding_kind * glob_constr * glob_constr -> unsafe_judgment pretype_fun;
+  pretype_prod : pretyper -> Name.t * Decl_kinds.binding_kind * glob_constr * glob_constr -> unsafe_judgment pretype_fun;
+  pretype_letin : pretyper -> Name.t * glob_constr * glob_constr option * glob_constr -> unsafe_judgment pretype_fun;
+  pretype_cases : pretyper -> Constr.case_style * glob_constr option * tomatch_tuples * cases_clauses -> unsafe_judgment pretype_fun;
+  pretype_lettuple : pretyper -> Name.t list * (Name.t * glob_constr option) * glob_constr * glob_constr -> unsafe_judgment pretype_fun;
+  pretype_if : pretyper -> glob_constr * (Name.t * glob_constr option) * glob_constr * glob_constr -> unsafe_judgment pretype_fun;
+  pretype_rec : pretyper -> fix_kind * Id.t array * glob_decl list array * glob_constr array * glob_constr array -> unsafe_judgment pretype_fun;
+  pretype_sort : pretyper -> glob_sort -> unsafe_judgment pretype_fun;
+  pretype_hole : pretyper -> Evar_kinds.t * intro_pattern_naming_expr * Genarg.glob_generic_argument option -> unsafe_judgment pretype_fun;
+  pretype_cast : pretyper -> glob_constr * glob_constr cast_type -> unsafe_judgment pretype_fun;
+  pretype_proj : pretyper -> Projection.t * glob_constr -> unsafe_judgment pretype_fun;
+  pretype_type : pretyper -> glob_constr -> bool -> constr option -> ExtraEnv.t -> evar_map ref -> int * ltac_var_map -> unsafe_type_judgment;
+}
+
+val default_pretyper : pretyper
+
+val eval_pretyper : pretyper -> glob_constr -> bool -> type_constraint -> ExtraEnv.t -> evar_map ref -> (int * ltac_var_map) -> unsafe_judgment
+val eval_type_pretyper : pretyper -> glob_constr -> bool -> constr option -> ExtraEnv.t -> evar_map ref -> (int * ltac_var_map) -> unsafe_type_judgment
```

Comment:
I think the second `eval` needs a comment, since from the signature I can hardly guess the difference...
I'm not even convinced that `type_constraint` makes no sense here (one could just say that IsType and NoConstraint are the same in the second API), or call the former with IsType and remove the second API.